### DISCUSSION TITLE
How to use msquic and webrtc at the same time?

### DISCUSSION THREAD

**Discussion by FH0** [microsoft/msquic#3471] [Q&A] [answered]
My application use [webrtc](https://webrtc.googlesource.com/src/) and msquic at the same time. It has webrtc directory, quic directory and main.cpp. I can't finish compiling because webrtc using boringssl and msquic using openssl, boringssl and openssl having many same export symbols.

I know `--exclude-libs ALL` flag. But it works for shared library. I need to compile my application statically.

**Comment by nibanks** [comment]
I don't know if there is any solution to try to statically link both boringssl and openssl into the same process. BoringSSL was forked from openssl some time ago, that's why they're so similar. Perhaps @wfurt has some ideas, but otherwise it's likely a question beyond the scope of this project.

**Reply by wfurt** [reply]
AFAIK you can compile libmsquic and link OpenSSL statically instead of using system crypto. Then it would not have the symbols exported. But then you cannot link your app statically. 
I think the only good solution IMHO is to either make webrtc work with OpenSSL or MsQuic with BoringSsl. Since they share same origin it may not be that hard.

**Comment by FH0** [comment, marked as answer]
Finally solved by `objcopy`. Only prefix with `msquic|quic` symbols are exported.

```makefile
compile_msquic: update_submodule
	mkdir -p ./dep/msquic/$(TARGET)
	sed -i 's|\(^ *msquic_lib\)$$|\1 ALL|g' ./dep/msquic/src/bin/CMakeLists.txt
	cmake -B./dep/msquic/$(TARGET) -S./dep/msquic -DQUIC_BUILD_SHARED=OFF -DCMAKE_TARGET_ARCHITECTURE=$(TARGET_CPU)
	make -C./dep/msquic/$(TARGET) -j$(shell nproc)
	$(eval renameSymbols=$(shell objdump -t ./dep/msquic/$(TARGET)/bin/Release/libmsquic.a | awk -v RS= '/_YB80VJ/{next}1' | grep -E 'g +F .text' | grep -Evi ' (ms){0,1}quic' | awk '{print " --redefine-sym " $$NF "=" $$NF "_YB80VJ"}'))
	@objcopy $(renameSymbols) ./dep/msquic/$(TARGET)/bin/Release/libmsquic.a
```
